File: scripts/update_data.R
# File: scripts/update_data.R
# Purpose: pull season-to-date MLB leaders from FanGraphs and overwrite
#          batters_2025_2025.csv and pitchers_2025_2025.csv in repo root.
# Notes:
# - Preserves your existing Salary (Per Season) by merging from current CSVs.
# - Drops aggregated rows like "2 Tms" / "3 Tms".
# - Uses FanGraphs CSV export with a friendly User-Agent.

suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
  library(stringr)
  library(httr)
})

# ---- FanGraphs URLs (season-to-date) ----
# You can adjust enddate/qual in these URLs as needed.
BATTERS_URL  <- "https://www.fangraphs.com/leaders/major-league?pageitems=2000000000&startdate=2025-03-27&enddate=2025-08-09&season=2025&season1=2025&ind=0&team=0&pos=all&type=8&stats=bat&qual=1&month=1000"
PITCHERS_URL <- "https://www.fangraphs.com/leaders/major-league?pageitems=2000000000&startdate=2025-03-27&enddate=2025-08-09&season=2025&season1=2025&ind=0&team=0&pos=all&type=8&stats=pit&qual=1&month=1000"

# ---- Helpers ----
safe_select <- function(df, cols) dplyr::select(df, dplyr::any_of(cols))

fetch_fg_csv <- function(url) {
  u <- if (grepl("csv=1", url, fixed = TRUE)) url else paste0(url, "&csv=1")
  resp <- httr::GET(u, httr::user_agent("mlb-value-tool (github.com/Martin9803/mlb-value-tool)"))
  httr::stop_for_status(resp)
  readr::read_csv(httr::content(resp, as = "raw"), show_col_types = FALSE)
}

merge_salary <- function(new_df, existing_path) {
  if (file.exists(existing_path)) {
    sal <- readr::read_csv(existing_path, show_col_types = FALSE) %>%
      safe_select(c("Name", "Team", "Salary (Per Season)")) %>%
      mutate(Name = str_squish(Name), Team = str_trim(Team))
    new_df <- new_df %>%
      mutate(Name = str_squish(Name)) %>%
      left_join(sal, by = c("Name", "Team"))
  } else {
    new_df <- new_df %>% mutate(`Salary (Per Season)` = NA_real_)
  }
  # Drop legacy index columns if they exist
  new_df %>% select(-any_of(c("Unnamed: 0", "Unnamed: 0.1")))
}

# ---- Download ----
message("Downloading FanGraphs leaderboards…")
bat_raw <- fetch_fg_csv(BATTERS_URL)
pit_raw <- fetch_fg_csv(PITCHERS_URL)

# ---- Normalize columns your app expects ----
# Batters
bat <- bat_raw %>%
  rename(Name = "Name", Team = "Team") %>%
  mutate(Team = str_trim(Team)) %>%
  filter(!Team %in% c("2 Tms", "3 Tms")) %>%
  safe_select(c(
    "Name", "Team", "G", "PA", "R", "RBI", "SB", "BB%", "K%",
    "AVG", "OBP", "SLG", "HR", "WAR"
  ))

# Pitchers
pit <- pit_raw %>%
  rename(Name = "Name", Team = "Team") %>%
  mutate(Team = str_trim(Team)) %>%
  filter(!Team %in% c("2 Tms", "3 Tms")) %>%
  safe_select(c(
    "Name", "Team", "G", "IP", "W", "L", "SV", "K/9", "BB/9", "HR/9",
    "ERA", "FIP", "WAR"
  ))

# ---- Merge salary from current CSVs ----
bat <- merge_salary(bat, "batters_2025_2025.csv")
pit <- merge_salary(pit, "pitchers_2025_2025.csv")

# ---- Write out ----
readr::write_csv(bat, "batters_2025_2025.csv")
readr::write_csv(pit, "pitchers_2025_2025.csv")

message(sprintf("Saved batters_2025_2025.csv (%s rows)", nrow(bat)))
message(sprintf("Saved pitchers_2025_2025.csv (%s rows)", nrow(pit)))

File: scripts/update_data.py
#!/usr/bin/env python3
"""
Pull season-to-date MLB leaders from FanGraphs and overwrite the two CSVs used by
your app: `batters_2025_2025.csv` and `pitchers_2025_2025.csv`.

- Uses FanGraphs CSV export (adds `csv=1`) with a friendly User-Agent
- Filters out "2 Tms/3 Tms"
- Selects/renames only the columns your app uses
- Preserves your `Salary (Per Season)` by merging from the existing CSVs
- Writes in the same schema/order your app expects
"""
from __future__ import annotations
import os
import sys
import time
from io import BytesIO
from datetime import date
import pandas as pd
import requests

BATTERS_URL  = (
    "https://www.fangraphs.com/leaders/major-league"
    "?pageitems=2000000000&startdate=2025-03-27&enddate={end}"
    "&season=2025&season1=2025&ind=0&team=0&pos=all&type=8&stats=bat&qual=1&month=1000"
)
PITCHERS_URL = (
    "https://www.fangraphs.com/leaders/major-league"
    "?pageitems=2000000000&startdate=2025-03-27&enddate={end}"
    "&season=2025&season1=2025&ind=0&team=0&pos=all&type=8&stats=pit&qual=1&month=1000"
)

HEADERS = {"User-Agent": "mlb-value-tool (github.com/Martin9803/mlb-value-tool)"}
BATTERS_OUT  = "batters_2025_2025.csv"
PITCHERS_OUT = "pitchers_2025_2025.csv"

# Columns your app expects, in order
BATTERS_ORDER = [
    "#", "Name", "Team", "Type", "PA", "G", "WAR", "SLG", "OBP", "HR",
    "RBI", "SB", "R", "BB%", "K%", "AVG", "Salary (Per Season)"
]
PITCHERS_ORDER = [
    "#", "Name", "Team", "Type", "W", "L", "SV", "G", "IP", "K/9", "BB/9",
    "HR/9", "ERA", "FIP", "WAR", "Salary (Per Season)"
]


def _today_str() -> str:
    return os.getenv("END_DATE", date.today().strftime("%Y-%m-%d"))


def fetch_fg_csv(url: str, retries: int = 3, backoff: float = 1.5) -> pd.DataFrame:
    """Fetch a FanGraphs CSV into a DataFrame with simple retry/backoff."""
    if "csv=1" not in url:
        url = f"{url}&csv=1"
    last_exc = None
    for i in range(retries):
        try:
            resp = requests.get(url, headers=HEADERS, timeout=60)
            resp.raise_for_status()
            return pd.read_csv(BytesIO(resp.content))
        except Exception as exc:  # noqa: BLE001 - log & retry
            last_exc = exc
            time.sleep(backoff ** (i + 1))
    raise RuntimeError(f"Failed to fetch FanGraphs CSV: {last_exc}")


def merge_salary(df: pd.DataFrame, existing_path: str) -> pd.DataFrame:
    """Preserve Salary (Per Season) by merging from existing CSV if present."""
    if os.path.exists(existing_path):
        existing = pd.read_csv(existing_path, encoding="ISO-8859-1")
        if "Salary (Per Season)" in existing.columns:
            sal = existing[["Name", "Team", "Salary (Per Season)"]].copy()
            sal["Name"] = sal["Name"].astype(str).str.strip()
            sal["Team"] = sal["Team"].astype(str).str.strip()
            df["Name"] = df["Name"].astype(str).str.strip()
            df["Team"] = df["Team"].astype(str).str.strip()
            df = df.merge(sal, on=["Name", "Team"], how="left")
            return df
    df["Salary (Per Season)"] = pd.NA
    return df


def normalize_batters(df: pd.DataFrame) -> pd.DataFrame:
    # Drop multi-team summary rows
    df = df[df["Team"].isin(["2 Tms", "3 Tms"]) == False].copy()
    # Map/keep only the columns we need
    keep = {
        "Name": "Name",
        "Team": "Team",
        "G": "G",
        "PA": "PA",
        "WAR": "WAR",
        "SLG": "SLG",
        "OBP": "OBP",
        "HR": "HR",
        "RBI": "RBI",
        "SB": "SB",
        "R": "R",
        "BB%": "BB%",
        "K%": "K%",
        "AVG": "AVG",
    }
    out = pd.DataFrame({new: df.get(old) for old, new in keep.items()})
    out.insert(0, "#", range(1, len(out) + 1))
    out.insert(3, "Type", "Batter")
    return out


def normalize_pitchers(df: pd.DataFrame) -> pd.DataFrame:
    df = df[df["Team"].isin(["2 Tms", "3 Tms"]) == False].copy()
    keep = {
        "Name": "Name",
        "Team": "Team",
        "W": "W",
        "L": "L",
        "SV": "SV",
        "G": "G",
        "IP": "IP",
        "K/9": "K/9",
        "BB/9": "BB/9",
        "HR/9": "HR/9",
        "ERA": "ERA",
        "FIP": "FIP",
        "WAR": "WAR",
    }
    out = pd.DataFrame({new: df.get(old) for old, new in keep.items()})
    out.insert(0, "#", range(1, len(out) + 1))
    out.insert(3, "Type", "Pitcher")
    return out


def write_ordered(df: pd.DataFrame, order: list[str], path: str) -> None:
    # Ensure all columns exist; missing -> NA
    for col in order:
        if col not in df.columns:
            df[col] = pd.NA
    df = df[order]
    # Remove legacy index columns if ever present
    for c in list(df.columns):
        if c.lower().startswith("unnamed:"):
            del df[c]
    df.to_csv(path, index=False)


def main() -> int:
    end = _today_str()
    print(f"Fetching FanGraphs leaders through {end}…", flush=True)

    bat_raw = fetch_fg_csv(BATTERS_URL.format(end=end))
    pit_raw = fetch_fg_csv(PITCHERS_URL.format(end=end))

    bat = normalize_batters(bat_raw)
    pit = normalize_pitchers(pit_raw)

    bat = merge_salary(bat, BATTERS_OUT)
    pit = merge_salary(pit, PITCHERS_OUT)

    # Final ordering & write
    write_ordered(bat, BATTERS_ORDER, BATTERS_OUT)
    write_ordered(pit, PITCHERS_ORDER, PITCHERS_OUT)

    print(f"Saved {BATTERS_OUT}: {len(bat)} rows")
    print(f"Saved {PITCHERS_OUT}: {len(pit)} rows")
    return 0


if __name__ == "__main__":
    sys.exit(main())
